src/adapters/suggest_sources.js

```diff
@@ -0,0 +1,55 @@
+import PoiStore from './poi/poi_store';
+import BragiPoi from './poi/bragi_poi';
+import CategoryService from './category_service';
+
+// @TODO: Improvement: don't access directly to window.map
+function getFocus(focusMinZoom) {
+  const zoom = window.map && window.map.mb && window.map.mb.getZoom();
+  if (zoom >= focusMinZoom) {
+    const { lat, lon } = window.map.mb.getCenter();
+    return { lat, lon, zoom };
+  }
+  return {};
+}
+
+export function suggestResults(term, { withCategories, useFocus, focusMinZoom = 11 } = {}) {
+  let geocoderPromise;
+  let promise;
+  if (term === '') {
+    // Prerender Favorites on focus in empty field
+    promise = PoiStore.getAll();
+  } else {
+    promise = new Promise(async (resolve, reject) => {
+      geocoderPromise = BragiPoi.get(term, useFocus ? getFocus(focusMinZoom) : {});
+      const favoritePromise = PoiStore.get(term);
+      const categoryPromise = withCategories ? CategoryService.getMatchingCategories(term) : [];
+
+      try {
+        const [geocoderSuggestions, favorites, categories] =
+          await Promise.all([ geocoderPromise, favoritePromise, categoryPromise ]);
+
+        // For now don't display anything when there is no geocoder match
+        if (geocoderSuggestions.length === 0) {
+          return resolve([]);
+        }
```

Comment:
This is actually not equivalent to the current implementation, although arguably it could have been made clearer :)

`geocoderSuggesstions` will be `null` if the geocoder query has been aborted (by successive keystrokes), and that's why the current implementation calls `resolve(null)` in that case to avoid refresh the autocomplete results with an obsolete request.

An empty list should not be caught here, as some suggestions might exist among categories of favorites.

Reply:
Thanks for the explanation! I had the wrong analysis and conclusion. I'll just restore the previous behavior and change the comment.